File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

    static int N, E;
    static int v1, v2;
    static ArrayList<Node>[] list;
    static int[] dist;
    static boolean[] visit;
    private static final int INF = 200000 * 1000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());
        list = new ArrayList[N + 1];
        dist = new int[N + 1];
        for (int i = 1; i <= N; i++) list[i] = new ArrayList<>();
        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            list[a].add(new Node(b, c));
            list[b].add(new Node(a, c));
        }
        st = new StringTokenizer(br.readLine());
        v1 = Integer.parseInt(st.nextToken());
        v2 = Integer.parseInt(st.nextToken());

        int ans1 = 0;
        ans1 += dijkstra(1, v1);
        ans1 += dijkstra(v1, v2);
        ans1 += dijkstra(v2, N);

        int ans2 = 0;
        ans2 += dijkstra(1, v2);
        ans2 += dijkstra(v2, v1);
        ans2 += dijkstra(v1, N);

        int ans = Math.min(ans1, ans2);
        System.out.println(ans >= INF ? -1 : ans);
    }

    private static int dijkstra(int start, int end) {
        Arrays.fill(dist, INF);
        visit = new boolean[N + 1];
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(start, 0));
        dist[start] = 0;

        while (!pq.isEmpty()) {
            Node cur = pq.poll();

            if (visit[cur.index]) continue;
            visit[cur.index] = true;

            for (Node next : list[cur.index]) {
                if (dist[next.index] > cur.cost + next.cost) {
                    dist[next.index] = cur.cost + next.cost;
                    pq.offer(new Node(next.index, dist[next.index]));
                }
            }
        }
        return dist[end];
    }

    static class Node implements Comparable<Node> {
        int index, cost;

        Node(int index, int cost) {
            this.index = index;
            this.cost = cost;
        }

        @Override
        public int compareTo(Node o) {
            return Integer.compare(cost, o.cost);
        }
    }
}

File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.kt
import java.util.*

const val INF = 200000 * 1000

fun main() {

    data class Node(val next: Int, val cost: Int) : Comparable<Node> {
        override fun compareTo(other: Node) = cost.compareTo(other.cost)
    }
    val (n, e) = readln().split(" ").map { it.toInt() }
    val list = Array(n + 1) { ArrayList<Node>() }
    for (i in 0 until e) {
        val (a, b, c) = readln().split(" ").map { it.toInt() }
        list[a].add(Node(b, c))
        list[b].add(Node(a, c))
    }
    val (need1, need2) = readln().split(" ").map { it.toInt() }

    fun dijkstra(start: Int, end: Int): Int {
        val dist = IntArray(n + 1) { INF }
        val pq = PriorityQueue<Node>()
        val visit = BooleanArray(n + 1)
        dist[start] = 0
        pq.offer(Node(start, dist[start]))

        while (!pq.isEmpty()) {
            val cur = pq.poll()
            if (visit[cur.next]) continue
            visit[cur.next] = true

            for (k in list[cur.next].indices) {
                val mn = list[cur.next][k]
                if (visit[mn.next]) continue
                if (dist[mn.next] > dist[cur.next] + mn.cost) {
                    dist[mn.next] = dist[cur.next] + mn.cost
                    pq.offer(Node(mn.next, dist[mn.next]))
                }
            }
        }
        return dist[end]
    }

    var ans1 = 0
    ans1 += dijkstra(1, need1)
    ans1 += dijkstra(need1, need2)
    ans1 += dijkstra(need2, n)

    var ans2 = 0
    ans2 += dijkstra(1, need2)
    ans2 += dijkstra(need2, need1)
    ans2 += dijkstra(need1, n)

    println(ans1.coerceAtMost(ans2).takeIf { it < INF } ?: -1)
}